_posts/2018-03-23-minnowboard-turbot-remote-firmware-flashing-with-remote-testing-environment.md

```diff
@@ -0,0 +1,312 @@
+
+* UART debug
+
+| Minnowboard J4 | Color         | UART-USB converter |
+|:--------------:|:-------------:|:------------------:|
+| 1              | black         | GND                |
+| 2              | -             | NC                 |
+| 3              | -             | NC                 |
+| 4              | blue          | TX                 |
+| 5              | yellow        | RX                 |
+| 6              | -             | NC                 |
+
+
+There was one but important problem with the built system. Memory flashing was
+realized correctly, but Minnowboard doesn't boot. The source of the problem 
+was `CS#` signal. It was set to `0`, so memory was always enabled, but it was 
+not enough. Minnowboard still doesn't boot. So I decided to use an 
+oscilloscope. That analysis showed me that Minnowboard Turbot doesn't set 
+memory chip's `CS#` input to `0` all the time. It chnges over the time. Mostly
+`CS#` is set to `0`, but periodically it's being set to `1`.
+
+Time analysis of the `CS#` states seemed to be too time-consuming and trying
+correct control of the `CS#` lines too inflexible in the event of changes.
+I needed to reduce the impact of the `CS#` RTE output to make it a Minnowboard
+Turbot took precedence in choosing the state of the `CS#` line.
+I decided to use 1.2 kΩ resistor between RTE `CS#` output and Minnowboard 
+Turbot `CS#` input and this was a good idea.
+
+After resolving `CS#` problem, flashing procedure is very simple.
+It amounts to: 
+1. Turn off Minnowboard platform power supply.
+2. Flash Minnowboard ROM memory via SPI.
+3. Turn on Minnowboard platform power supply.
+
+And it's that all.
+
+### Firmware flashing
+
+I used 2 firmwares to test presented Remote Testing Environment:
+
+[MNW2MAX1.X64.0097.D01.1709211100.bin](https://firmware.intel.com/sites/default/files/MinnowBoard_MAX-0.96-Firmware.Images.zip),
+
+[MNW2MAX1.X64.0097.D01.1709211100.bin](https://firmware.intel.com/sites/default/files/MinnowBoard_MAX-Rel_0_97-Firmware.Images.zip).
+
+
+Code of flashing script:
+```
+#!/bin/bash
+
+# turn off Minnowboard platform power supply
+../rte_ctrl -rel
+
+sleep 1
+
+# flash Minnowboard ROM with a new firmware
+flashrom -p linux_spi:dev=/dev/spidev1.0,spispeed=32000 -w $1
+
+sleep 1
+
+# turn on Minnowboard platform power supply
+../rte_ctrl -rel
+```
+
+> `rte_ctrl` is a program written to control power supply of DUT. For the use 
+shown in this article it can be fully replaced by direct output control of GPIO
+which is being used to turn on and turn off power supply of DUT. In this
+situation flashing script will look like this:
+
+```
+#!/bin/bash
+
+# turn off Minnowboard platform power supply
+echo 0 > /sys/class/gpio/gpio199/value
+
+sleep 1
+
+# flash Minnowboard ROM with a new firmware
+flashrom -p linux_spi:dev=/dev/spidev1.0,spispeed=32000 -w $1
+
+sleep 1
+
+# turn on Minnowboard platform power supply
+echo 1 > /sys/class/gpio/gpio199/value
+```
```

Comment:
@acihy in light of this instruction maybe we even don't need info about rte_ctrl - I believe if we decide info about rte_ctrl should be removed - we should keep the content you wrote somewhere in rte_ctrl or RTE documentation.